test/helpers/envtest.go

```diff
@@ -308,21 +310,88 @@ func (t *TestEnvironment) Cleanup(ctx context.Context, objs ...client.Object) er
 	for _, o := range objs {
 		err := t.Client.Delete(ctx, o)
 		if apierrors.IsNotFound(err) {
-			// If the object is not found, it must've been garbage collected
-			// already. For example, if we delete namespace first and then
-			// objects within it.
 			continue
 		}
 		errs = append(errs, err)
 	}
 	return kerrors.NewAggregate(errs)
 }
 
+// CleanupAndWait deletes all the given objects and waits for the cache to be updated accordingly.
+//
+// NOTE: Waiting for the cache to be updated helps in preventing test flakes due to the cache sync delays.
+func (t *TestEnvironment) CleanupAndWait(ctx context.Context, objs ...client.Object) error {
```

Comment:
Should we get Cleanup to always wait instead of having a different method?

Reply:
I initially went down this path but it seems tricky because e.g. namespaces does not get deleted in testenv (they have a finaliser an we are not running cube-controller-manager). 
Also this is really required only when you have tests deleting and recreating objects with the same name like kubeadm-config Map, so I decided to make this opt in and let the developer to choose when to wait for deletion.